### PR TITLE
Viewmanager.py: add check disk space

### PR DESCRIPTION
**Pull request by PushKK**
For https://gramps-project.org/bugs/view.php?id=12306
I add check disk space before Quit and BackUp.

Test this code please.

### CODE REVIEW

**Comment by tosky** [comment]
This looks like a duplicate of https://github.com/gramps-project/gramps/pull/1211 - wrong branch?

**Comment by PushKK** [comment]
Sorry, GitHub add all my previous PRs to last PR. Alltime.
For resolve this problem I delete my fork and create fork again...

**Comment by Nick-Hall** [comment]
@PushKK You probably just need to rebase your branches on _master_.

File: gramps/gui/viewmanager.py [hunk 1]
```diff
@@ -580,10 +580,47 @@ def no_del_event(self, *obj):
         hits 'x' multiple times. """
         return True
 
+    def checkdisk(self, dbfolder): # TODO check this function
```
Comment (Nick-Hall) [inline gramps/gui/viewmanager.py:583]: This belongs in [`gramps/gen/utils/file.py`](https://github.com/gramps-project/gramps/blob/master/gramps/gen/utils/file.py).

File: gramps/gui/viewmanager.py [hunk 2]
```diff
@@ -580,10 +580,47 @@ def no_del_event(self, *obj):
...
+        """
+        Check disk space
+        """
+        if win():                               # for windows:
```
Comment (Nick-Hall) [inline gramps/gui/viewmanager.py:587]: What should we do for Windows?

File: gramps/gui/viewmanager.py [hunk 3]
```diff
@@ -580,10 +580,47 @@ def no_del_event(self, *obj):
...
+        #                   - run (Gramps need space for load config file)
+        #                   - quit (Gramps need some space for:
+        #                                                   - db.commit and close database,
+        #                                                   - save config file)
+        #                   - every backup (we need space 'last_backup_file_size'+ 5%)
+        diskspace = 20             # Disk space for Linux - 20 MB, for Gramps - 5 MB, for backup 'last_backup_file_size' + 5%
+        if (self.checkdisk(config.get('database.path')) < diskspace):
+            self.uistate.push_message(self.dbstate, _("Disk space < " + str(diskspace) + " MB. Quit impossible. Clear your disk and repeat this operation."))
+            WarningDialog(
+                _("Message for user:"),
```
Comment (Nick-Hall) [inline gramps/gui/viewmanager.py:615]: Perhaps "Low disk space" instead?

File: gramps/gui/viewmanager.py [hunk 4]
```diff
@@ -580,10 +580,47 @@ def no_del_event(self, *obj):
...
+#****************
+        # We must check disk space before:
+        #                   - run (Gramps need space for load config file)
+        #                   - quit (Gramps need some space for:
+        #                                                   - db.commit and close database,
+        #                                                   - save config file)
+        #                   - every backup (we need space 'last_backup_file_size'+ 5%)
+        diskspace = 20             # Disk space for Linux - 20 MB, for Gramps - 5 MB, for backup 'last_backup_file_size' + 5%
+        if (self.checkdisk(config.get('database.path')) < diskspace):
+            self.uistate.push_message(self.dbstate, _("Disk space < " + str(diskspace) + " MB. Quit impossible. Clear your disk and repeat this operation."))
```
Comment (Nick-Hall) [inline gramps/gui/viewmanager.py:613]: Use format strings instead of concatenation.

File: gramps/gui/viewmanager.py [hunk 5]
```diff
@@ -1188,6 +1225,35 @@ def autobackup(self):
...
+
+        backup_name = self.dbstate.db.get_dbname()
+        backup_ext = ".gramps"
+        path = config.get('database.backup-path')
+        bpath = sorted(os.listdir(path))
+        for n, bfile in enumerate(bpath):
+            if os.path.isfile(os.path.join(path, bfile)) and bfile.startswith(backup_name) and bfile.endswith(backup_ext):
+                blastfile = bfile
+
+        bfile_size = round(os.path.getsize(os.path.join(path, blastfile))/(1024))
```
Comment (Nick-Hall) [inline gramps/gui/viewmanager.py:1241]: A utility function to return the largest backup size would be neater.

**Review by Nick-Hall — CHANGES_REQUESTED** [review]
Good idea.  It just needs a little refinement.

File: gramps/gui/viewmanager.py (hunk 2 above)
Comment (PushKK) [reply gramps/gui/viewmanager.py:587]: Gramps work in Windows too. it's draft for Windows. I check this code in my Linux system, but I have not Windows.

File: gramps/gui/viewmanager.py (hunk 3 above)
Comment (PushKK) [reply gramps/gui/viewmanager.py:615]: Yes, of course.

File: gramps/gui/viewmanager.py (hunk 4 above)
Comment (PushKK) [reply gramps/gui/viewmanager.py:613]: I'm not Python man.

File: gramps/gui/viewmanager.py (hunk 5 above)
Comment (PushKK) [reply gramps/gui/viewmanager.py:1241]: I think about this - we must use biggest or last backup file.
In result - Gramps must use last backup file for active database.
Reason:
If user delete non used info from database -> new_database_size < old_database_size.
And Gramps create new_database_backup with small size.

Comment (Nick-Hall) [reply gramps/gui/viewmanager.py:1241]: Yes.  It just needs to be put into a utility function in `gramps/gen/utils/file.py`.

File: gramps/gui/viewmanager.py (hunk 1 above)
Comment (PushKK) [reply gramps/gui/viewmanager.py:583]: Yes, in new PR this function move to file.py.

File: gramps/gui/viewmanager.py (hunk 4 above)
Comment (PushKK) [reply gramps/gui/viewmanager.py:613]: I rewrite this code. New PR have normal string concatenation.
I check code, all translated strings are present in .po file.

File: gramps/gui/viewmanager.py (hunk 5 above)
Comment (PushKK) [reply gramps/gui/viewmanager.py:1241]: I test this code two day.
In result - backup_file_size must be calculated from active_database_size.
For SQLite3 - new database with Test71 MB have 4,1 MB backup file (> 7%).
If user create new Family tree and import to tree info from other base - Gramps must say to user:
1. You need least 'calculated backup_file_size' MB
or (easy way on 'try - except')
2. You cannot save your file. Clear your disk and try save again.

**Comment by Nick-Hall** [comment]
Replaced by PR #1218.

**Comment by Nick-Hall** [comment]
Replaced by PR #1218.

**Pull request closed without merging**